### PR TITLE
Add `SIMPLE` pagination type

### PR DESCRIPTION
**Pull request by thekonz**
- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

**Changes**

I noticed that we don't have a way to use `simplePaginate`. Sometimes I don't want to fire two queries. That can have multiple reasons. Response times of slower queries for example, or if you want to build an endless scrolling system you don't need to know the total number of items.

So I added a pagination type `SIMPLE`.

### CODE REVIEW

**Comment by thekonz** [comment]
I sprinkled in some tests, but I am not sure if you want some more. Happy to add more if needed. I personally feel like they're enough.

File: docs/master/api-reference/directives.md [hunk 1]
```diff
@@ -2156,6 +2178,64 @@ type PostEdge {
 }
 ```
 
+### Simple Pagination
+
+The default `PAGINATOR` uses the query builder method `paginate` which will
```
Comment (spawnia) [inline docs/master/api-reference/directives.md:2183]: Is there a better place for this explanatory text? I find it a bit confusing to have this whole paragraph about another pagination method as the first thing to read under `Simple Pagination`.

Comment (spawnia) [reply docs/master/api-reference/directives.md:2183]:
```suggestion
In contrast to other pagination types, `SIMPLE` pagination only fires a single database
query on every request. This improves performance, but means that the response does not
hold information about the total number of items.
```

File: docs/master/api-reference/directives.md [hunk 2]
```diff
@@ -2156,6 +2178,64 @@ type PostEdge {
...
+  ): PostSimplePaginator
+}
+
+"A paginated list of Post items."
+type PostSimplePaginator {
+  "A list of Post items."
+  data: [Post!]!
+
+  "Pagination information about the list of items."
+  paginatorInfo: SimplePaginatorInfo!
```
Comment (spawnia) [inline docs/master/api-reference/directives.md:2219]: Let's include the definition for `SimplePaginatorInfo` in the docs as well. If the other types are lacking that information, we should add it.

File: src/Pagination/PaginationArgs.php [hunk 3]
```diff
@@ -87,13 +95,15 @@ protected static function calculateCurrentPage(int $first, int $after, int $defa
      *
      * @param  \Illuminate\Database\Query\Builder|\Laravel\Scout\Builder|\Illuminate\Database\Eloquent\Builder|\Illuminate\Database\Eloquent\Relations\Relation  $builder
      */
-    public function applyToBuilder(object $builder): LengthAwarePaginator
+    public function applyToBuilder(object $builder): Paginator
     {
+        $methodName = $this->type->isSimple() ? 'simplePaginate' : 'paginate';
```
Comment (spawnia) [inline src/Pagination/PaginationArgs.php:100]:
```suggestion
        $methodName = $this->type->isSimple()
            ? 'simplePaginate'
            : 'paginate';
```

File: src/Pagination/PaginationServiceProvider.php [hunk 4]
```diff
@@ -62,6 +63,32 @@ protected static function paginatorInfo(): ObjectTypeDefinitionNode
...
+              count: Int!
+
+              "Current pagination page."
+              currentPage: Int!
+
+              "Index of first item in the current page."
+              firstItem: Int
+
+              "If collection has more pages."
+              hasMorePages: Boolean!
```
Comment (spawnia) [inline src/Pagination/PaginationServiceProvider.php:81]: Is that really available?

File: tests/Integration/Pagination/PaginateDirectiveDBTest.php [hunk 5]
```diff
@@ -470,4 +471,92 @@ public function testDoesNotRequireDefaultCountArgIfDefinedInConfig(): void
...
+
+        DB::enableQueryLog();
+        $this->schema = /** @lang GraphQL */ '
+        type User {
+            id: ID!
+            name: String!
+        }
+
+        type Query {
+            users: [User!] @paginate(type: PAGINATOR)
```
Comment (spawnia) [inline tests/Integration/Pagination/PaginateDirectiveDBTest.php:488]:
```suggestion
            usersPaginator: [User!] @paginate(type: PAGINATOR)
```

File: tests/Integration/Pagination/PaginateDirectiveDBTest.php [hunk 6]
```diff
@@ -470,4 +471,92 @@ public function testDoesNotRequireDefaultCountArgIfDefinedInConfig(): void
...
+            users2: [User!] @paginate(type: SIMPLE)
```
Comment (spawnia) [inline tests/Integration/Pagination/PaginateDirectiveDBTest.php:489]:
```suggestion
            usersSimple: [User!] @paginate(type: SIMPLE)
```

File: tests/Integration/Schema/Directives/MorphManyDirectiveTest.php [hunk 7]
```diff
@@ -175,6 +175,7 @@ public function testQueryMorphManyPaginator(): void
             id: ID!
             title: String!
             images: [Image!] @morphMany(type: PAGINATOR)
+            images2: [Image!] @morphMany(type: SIMPLE, relation: "images")
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/MorphManyDirectiveTest.php:178]:
```suggestion
            imagesSimple: [Image!] @morphMany(type: SIMPLE, relation: "images")
```

File: tests/Integration/Schema/Directives/MorphManyDirectiveTest.php [hunk 8]
```diff
@@ -175,6 +175,7 @@ public function testQueryMorphManyPaginator(): void
             id: ID!
             title: String!
             images: [Image!] @morphMany(type: PAGINATOR)
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/MorphManyDirectiveTest.php:177]:
```suggestion
            imagesPaginator: [Image!] @morphMany(type: PAGINATOR)
```

File: tests/Unit/Pagination/PaginateDirectiveTest.php [hunk 9]
```diff
@@ -204,6 +234,7 @@ public function testIsLimitedToMaxCountFromConfig(): void
         type Query {
             users1: [User!]! @paginate
             users2: [User!]! @paginate(type: CONNECTION)
+            users3: [User!]! @paginate(type: SIMPLE)
```
Comment (spawnia) [inline tests/Unit/Pagination/PaginateDirectiveTest.php:237]: Let's switch to meaningful field names here too.

File: tests/Unit/Pagination/PaginateDirectiveTest.php [hunk 10]
```diff
@@ -52,6 +52,7 @@ public function testOnlyRegistersOneTypeForMultiplePaginators(): void
             users: [User!]! @paginate
             users2: [User!]! @paginate(type: CONNECTION)
             users3: [User!]! @paginate(type: "relay")
+            users4: [User!]! @paginate(type: "simple")
```
Comment (spawnia) [inline tests/Unit/Pagination/PaginateDirectiveTest.php:55]: Let's switch to meaningful field names here too.

**Review by spawnia — COMMENTED** [review]
Looks pretty good overall.

Agreed on the extent of testing, that part looks good. I would like some cleanup to balance the maintenance burden of this new feature but am happy to include it. It seems pretty useful.

**Comment by spawnia** [comment]
Oh and it looks like older Laravel versions are causing problems: https://github.com/nuwave/lighthouse/pull/1816/checks?check_run_id=2392547721. If you can make it compatible that would be great, but I am also fine with including this as a Laravel 6+ only feature.

File: docs/master/api-reference/directives.md (hunk 1 above)
Comment (thekonz) [reply docs/master/api-reference/directives.md:2183]: Oh yes, that's true. Very good 👍

**Comment by canatufkansu** [comment]
Hi, is there any plan to merge this request soon ?

**Review by spawnia — APPROVED** [review]
Nicely done

**Comment by tlaverdure** [comment]
Great addition! I've been meaning to look into this. 🙌🏾

**Pull request merged**